asf_search/search/search_generator.py

```diff
@@ -78,78 +78,78 @@ def search_generator(
-                    last_page = ASFSearchResults(hits[:min(maxResults - count, len(hits))], opts=opts)
-                    count += len(last_page)
-                    # results.extend(hits[:min(maxResults - len(results), len(hits))])
-                    if count == maxResults:
-                        last_page.searchComplete = True
-                        yield last_page
-                        return
-                    else:
-                        yield last_page
+            
+            opts.session.headers.update({'CMR-Search-After': cmr_search_after_header})
+            total, last_page = process_page(items, maxResults, subquery_max_results, total, subquery_count, opts)
+            subquery_count += len(last_page)
+            last_page.searchComplete = subquery_count == subquery_max_results or total == maxResults
+            yield last_page
+            
+            if last_page.searchComplete:
+                if subquery_count == subquery_max_results and maxResults is None: # end of subquery
+                    cmr_search_after_header = None
+                elif total == maxResults:
+                    opts.session.headers.pop('CMR-Search-After', None)
+                    return
                 else:
-                    count += len(hits)
-                    yield ASFSearchResults(hits, opts=opts)
-
+                    cmr_search_after_header = None
+        
         opts.session.headers.pop('CMR-Search-After', None)
 
 
 @retry(reraise=True,
-       retry=retry_if_exception_type((TimeoutError, ASFSearch5xxError)),
-       wait=wait_exponential(multiplier=1, min=4, max=10),
-       stop=stop_after_delay(340),
+       retry=retry_if_exception_type(CMRIncompleteError),
+       wait=wait_fixed(2),
+       stop=stop_after_delay(120),
+    )
+def query_cmr(session: ASFSession, url: str, translated_opts: dict, sub_query_count: int):
+    response = get_page(session=session, url=url, translated_opts=translated_opts)
+
+    items = [ASFProduct(f, session=session) for f in response.json()['items']]
+    hits: int = response.json()['hits'] # total count of products given search opts
+
+    # sometimes CMR returns results with the wrong page size
+    if len(items) != INTERNAL.CMR_PAGE_SIZE and len(items) + sub_query_count < hits:
+        raise CMRIncompleteError(f"CMR returned page of incomplete results. Expected {min(INTERNAL.CMR_PAGE_SIZE, hits - sub_query_count)} results, got {len(items)}")
+
+    return items, hits, response.headers.get('CMR-Search-After', None)
+    
+
+def process_page(items: list[ASFProduct], max_results: int, subquery_max_results: int, total: int, subquery_count: int, opts: ASFSearchOptions):
+    if max_results is None:
+        last_page = ASFSearchResults(items[:min(subquery_max_results - subquery_count, len(items))], opts=opts)
+    else:
+        last_page = ASFSearchResults(items[:min(max_results - total, len(items))], opts=opts)
+    total += len(last_page)
+
+    return total, last_page
```

Comment:
Since all `total` does here is add the length of the page to a param, is it worth making the methods that call this do the logic instead? It'd take out one param/return value in this method. Returning one param is clean code in c++, but I forget if python has the same rules 😅.

Is `process_page` called in enough areas to make the extra return value worth it?

Reply:
And if you happen to combine the two comments above, you can just return the page directly:
```python
if max_results is None:
    num_results = subquery_max_results - subquery_count
else:
    num_results = max_results - total
return ASFSearchResults(num_results, opts=opts)
```